### ISSUE TITLE
Get device type in InferType

### ISSUE THREAD

**Issue by Godricly** [dmlc/mxnet#2653]
Is it possible to get device type in InferType? Or can I use a macro to generate different compiling path? 
In cudnn fp16 batchnorm, mean and variance are actually stored as float type. So I'm countering different device data type issue now. 
if I use this [macro](https://github.com/dmlc/mxnet/blob/master/src/operator/batch_norm.cu#L15), is the CPU computation disabled? Or it is still able to run but with wrong data type for CPU.

**Comment by piiswrong** [comment]
This macro won't work in opproperty if that's what you are asking.
I probably will be easier to make the mshadow version also use float for mean and variance.

**Comment by Godricly** [comment]
If you try solving it in that way. U need to make mshadow operators able to take in different input type and to decide output tensor type. Also This is kind of breaking the definition of DType operator in MxNet.

**Comment by piiswrong** [comment]
Infer type already support that. You can assign any type to input/output if they are -1

**Comment by Godricly** [comment]
Yes, you can decide whatever the type you like in infertype, but does mxnet support stuffs like +-*/ between different types?

**Comment by vchuravy** [comment]
I think that you will need to explicitly cast them.
